### PR TITLE
parallel repair request decode

### PR DESCRIPTION
**Pull request by jbiseda**
#### Problem
Prioritizing repair requests requires oversampling packets beyond the per-iteration processing limit. Repair requests decode is not parallelized.

#### Summary of Changes
parallelize repair request decode using:
- 4 threads
- min packets per thread: 128

Fixes #

### CODE REVIEW

File: core/src/serve_repair.rs [hunk 1]
```diff
@@ -445,9 +464,103 @@ impl ServeRepair {
...
+        my_id: &Pubkey,
+        socket_addr_space: &SocketAddrSpace,
+        stats: &mut ServeRepairStats,
+        cluster_type: ClusterType,
+    ) -> Vec<RepairRequestWithMeta> {
+        thread_pool
+            .install(|| {
+                reqs_v
+                    .par_iter()
+                    .with_min_len(VERIFY_MIN_PACKETS_PER_THREAD)
```
Comment (behzadnouri) [inline core/src/serve_repair.rs:481]: Per my observations, `.with_min_len` still adds overhead if total number of packets is smaller than this min-len.
Might be better not to use thread-pool at all in that case; something like:
https://github.com/solana-labs/solana/blob/f24e30e4f/ledger/src/blockstore.rs#L2803-L2824

File: core/src/serve_repair.rs [hunk 2]
```diff
@@ -445,9 +464,103 @@ impl ServeRepair {
         }
     }
 
+    fn decode_requests(
+        thread_pool: &ThreadPool,
+        reqs_v: Vec<PacketBatch>,
+        epoch_staked_nodes: &Option<Arc<HashMap<Pubkey, u64>>>,
+        whitelist: &RwLockReadGuard<HashSet<Pubkey>>,
```
Comment (behzadnouri) [inline core/src/serve_repair.rs:471]: `RwLockReadGuard` is redundant. This can just be `&HashSet<Pubkey>`.

File: core/src/serve_repair.rs [hunk 3]
```diff
@@ -445,9 +464,103 @@ impl ServeRepair {
...
+                                    }
+                                    Error::RepairVerify(RepairVerifyError::SigVerify) => {
+                                        stats.err_sig_verify.fetch_add(1, Ordering::Relaxed);
+                                    }
+                                    Error::RepairVerify(RepairVerifyError::TimeSkew) => {
+                                        stats.err_time_skew.fetch_add(1, Ordering::Relaxed);
+                                    }
+                                    Error::RepairVerify(RepairVerifyError::Unsigned) => {
+                                        stats.err_unsigned.fetch_add(1, Ordering::Relaxed);
+                                    }
```
Comment (behzadnouri) [inline core/src/serve_repair.rs:516]: why is it better to increment these err counters here as opposed to inside `verify_signed_packet` as before?

File: core/src/serve_repair.rs [hunk 4]
```diff
@@ -445,9 +464,103 @@ impl ServeRepair {
...
+    ) -> Vec<RepairRequestWithMeta> {
+        thread_pool
+            .install(|| {
+                reqs_v
+                    .par_iter()
+                    .with_min_len(VERIFY_MIN_PACKETS_PER_THREAD)
+                    .flatten()
+                    .filter_map(|packet| {
+                        let request: RepairProtocol = match packet.deserialize_slice(..) {
+                            Ok(request) => request,
```
Comment (behzadnouri) [inline core/src/serve_repair.rs:485]: maybe define this `|packet| {` as a separate function or lambda to avoid extensive indentation below.

**Review by behzadnouri — COMMENTED** [review]
Repair traffic is pretty low compared to turbine, gossip, and tpu at high load time.
So the downside here is that we are potentially taking resources from those and allocate to repair.
For example, if repair requests traffic is < 2% of turbine, and we are using 4 threads here, then turbine/shred-sigverify would need 196 threads to keep load-per-thread proportionally balanced.

Or the other way around, if turbine/shred-sigverify is using < 48 threads, then one thread for repair is already too many.

From my perspective, we should keep repair load pretty low, and if we are observing high repair load or spikes in repair, instead of allocating more resources to repair, debug what is causing high repair and fix that.

**Comment by jbiseda** [comment]
> From my perspective, we should keep repair load pretty low, and if we are observing high repair load or spikes in repair, instead of allocating more resources to repair, debug what is causing high repair and fix that.

I agree that reducing the number of repairs is ideal, but decode cost has increased with the signed requests and oversampling. I've limited this change to testnet.

**Comment by behzadnouri** [comment]
> I agree that reducing the number of repairs is ideal, but decode cost has increased with the signed requests and oversampling. I've limited this change to testnet.

It is true that repair-service is doing more work than before, but gossip and turbine are already doing signature verifications with orders of magnitude more packets. So on a relative basis, even with a single thread, repair service has already more resources than it should.

Also with the patch referenced in https://github.com/solana-labs/solana/issues/28638 I am expecting repair load to go further down. https://github.com/solana-labs/solana/issues/28637 is another issue which once resolved can further decrease repair load.

**Comment by jbiseda** [comment]
closing, will track decode vs processing time with #29940.

**Pull request closed without merging**